lib/web/websocket/stream/websocketstream.js

```diff
@@ -0,0 +1,494 @@
+        }
+      }, { once: true })
+    }
+
+    // 9.  Let client be this 's relevant settings object .
+    const client = environmentSettingsObject.settingsObject
+
+    // 10. Run this step in parallel :
+    // 10.1. Establish a WebSocket connection given urlRecord , protocols , and client . [FETCH]
+    this.#controller = establishWebSocketConnection(
+      urlRecord,
+      protocols,
+      client,
+      this.#handler,
+      options
+    )
+  }
+
+  // The url getter steps are to return this 's url , serialized .
+  get url () {
+    return this.#url.toString()
+  }
+
+  // The opened getter steps are to return this 's opened promise .
+  get opened () {
+    return this.#openedPromise.promise
+  }
+
+  // The closed getter steps are to return this 's closed promise .
+  get closed () {
+    return this.#closedPromise.promise
+  }
+
+  // The close( closeInfo ) method steps are:
+  close (closeInfo = undefined) {
+    if (closeInfo !== null) {
+      closeInfo = webidl.converters.WebSocketCloseInfo(closeInfo)
+    }
+
+    // 1. Let code be closeInfo [" closeCode "] if present, or null otherwise.
+    const code = closeInfo.closeCode ?? null
+
+    // 2. Let reason be closeInfo [" reason "].
+    const reason = closeInfo.reason
+
+    // 3. Close the WebSocket with this , code , and reason .
+    closeWebSocketConnection(this.#handler, code, reason, true)
+  }
+
+  #write (chunk) {
+    // 1. Let promise be a new promise created in stream ’s relevant realm .
+    const promise = createDeferredPromise()
+
+    // 2. Let data be null.
+    let data = null
+
+    // 3. Let opcode be null.
+    let opcode = null
+
+    // 4. If chunk is a BufferSource ,
+    if (ArrayBuffer.isView(chunk) || types.isAnyArrayBuffer(chunk)) {
+      // 4.1. Set data to a copy of the bytes given chunk .
+      data = Buffer.from(chunk)
+
+      // 4.2. Set opcode to a binary frame opcode.
+      opcode = opcodes.BINARY
+    } else {
+      // 5. Otherwise,
+
+      // 5.1. Let string be the result of converting chunk to an IDL USVString .
+      //    If this throws an exception, return a promise rejected with the exception.
+      let string
+
+      try {
+        string = webidl.converters.USVString(chunk)
+      } catch (e) {
+        return Promise.reject(e)
+      }
+
+      // 5.2. Set data to the result of UTF-8 encoding string .
+      data = new TextEncoder().encode(string)
+
+      // 5.3. Set opcode to a text frame opcode.
+      opcode = opcodes.TEXT
+    }
+
+    // 6. In parallel,
+    // 6.1. Wait until there is sufficient buffer space in stream to send the message.
+
+    // 6.2. If the closing handshake has not yet started , Send a WebSocket Message to stream comprised of data using opcode .
+    if (this.#handler.closeState === sentCloseFrameState.NOT_SENT) {
+      const frame = new WebsocketFrameSend(data)
+
+      this.#handler.socket.write(frame.createFrame(opcode), () => {
+        promise.resolve(undefined)
+      })
```

Comment:
Why don't you use SendQueue?

Reply:
No need to, you can only write BufferSources and strings